### PR TITLE
Fix implementation of _parse_decimal_100() in AP_GPS_NMEA

### PR DESCRIPTION
**Pull request by Regelink**
_parse_decimal_100() did not parse the fractional part for negative numbers. Furthermore, use the third decimal (when present) for proper rounding.

### CODE REVIEW

**Comment by WickedShell** [comment]
I admit I looked briefly at this this morning but I haven't tried it (I
can't come up with anyway around here to generate a negative number to test
this against. (I can only think of negative altitude by flying below sea
level. Maybe SITL can do a NMEA gps?

**Comment by Regelink** [comment]
I noticed this bug when I was inspecting my log files after a flight (the area where I test is in fact below sea level). Before I committed the fix, I did a sanity check by copying the old version and new version of the _parse_decimal_100() function into a separate C file and calling it with _term filled with a floating point number in ASCII (I tried both negative, positive numbers with one, two or three decimals). The result was as expected (flaw and fix confirmed).

File: libraries/AP_GPS/AP_GPS_NMEA.cpp [hunk 1]
```diff
@@ -361,7 +429,21 @@ bool AP_GPS_NMEA::_term_complete()
...
+        case _GPS_SENTENCE_GPRMC + 12: // Mode indicator (RMC)
+            /* A = Autonomous mode
+             * D = Differential mode
+             * F = Float RTK  (specific for NV08C (not official NMEA 0813))
+             * R = Real Time Kinematic (specific for NV08C (not official NMEA 0813))
+             * E = Estimated (dead reckoning) mode
+             * M = Manual input mode
+             * S = Simulator mode
+             * N = Data not valid
+             */
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NMEA.cpp]: What GPS gave you this table? It doesn't match the documentation I can find for the GPRMC message anywhere. According to everything I've found the only standard is A=Data Valid or V=Data not valid (which you don't have listed here)

**Comment by WickedShell** [comment]
I haven't looked extensively at the _parse_decimal_100 yet, at first blush it looks fine, but I really would want to test that in SITL, which I'm not sure how to do. I'll try and get some guidance about making SITL provide other GPS inputs to test that.

**Comment by lucasdemarchi** [comment]
My suggestion would be to move this function as a static function in Util. Then you write a test case in the unit test like https://github.com/diydrones/ardupilot/blob/master/libraries/AP_HAL/utility/tests/test_own_ptr.cpp and https://github.com/diydrones/ardupilot/blob/master/libraries/AP_Math/tests/test_vectors.cpp

**Comment by jaluna** [comment]
We have noticed this bug recently. The altitude only shows integer numbers for negative values. See screenshot https://www.dropbox.com/s/od5lm86fhy2etd6/image007.png?dl=0

This version will fix the problem and also parse any number of decimal places.

int32_t AP_GPS_NORTH::_parse_decimal_100()
{	
    char *q, *p = _term;
		float frac = 0;
		int sign = ((*p == '-')? -1: 1);
    int32_t ret = 100L * atol(p);
    while (isdigit(*p) || *p == '-')
        ++p;
    
    // convert fractional part
    if (*p == '.') {
        q = p + 1;
        float frac_scale = 0.1f;
        while (*q && isdigit(*q)) {
            frac += DIGIT_TO_VAL(*q) * frac_scale;
            q++;
            frac_scale *= 0.1f;
        }
    }
    return ret+sign*round(frac*100L);
}

**Comment by lucasdemarchi** [comment]
oh... sorry, looks like I forgot to submit a PR with an working version. I will do that today

File: libraries/AP_GPS/AP_GPS_NMEA.cpp (hunk 1 above)
Comment (OXINARF) [inline libraries/AP_GPS/AP_GPS_NMEA.cpp]: It seems it was added in version 2.3: http://www.gpsinformation.org/dale/nmea.htm#2.3
But the _gps_data_good below is wrong, the E shouldn't be there according to the protocol. In reality I'm not sure we should even be setting _gps_data_good here, it is already checked before.

**Comment by lucasdemarchi** [comment]
New version: #4166

**Pull request closed without merging**